Sources/MongoSwift/Operations/Operation.swift

```diff
@@ -21,9 +21,14 @@ internal struct DefaultOperationExecutor: OperationExecutor {
     internal func execute<T: Operation>(_ operation: T,
                                         client: MongoClient,
                                         session: ClientSession?) throws -> T.OperationResult {
-        // TODO SWIFT-374: if session is non-nil, use its underlying Connection
+        // if a session was provided, use its underlying connection
+        if let session = session {
+            let conn = try session.getConnection(forUseWith: client)
+            return try operation.execute(using: conn, session: session)
+        }
+        // otherwise use a new connection from the pool
         return try client.connectionPool.withConnection { conn in
```

Comment:
We may run into trouble here with change streams since `watch` both performs I/O during creation of the stream _and_ holds onto the connection for future use. Currently this isn't a problem for us because cursors aren't constructed via operations.

Reply:
yeah... I think like cursors, change stream operations are probably not going to be executable via the executor.

Reply:
I think we probably want to avoid a scenario where certain operations aren't run through the executor. Cursors are okay to not Operation-ify imo since they're not actually performing any I/O. For change streams we'd want them to be actual `Operation`s.

Perhaps modifying the `Operation` protocol to include some computed property that determines whether the connection is still being held or not would be useful.

something like:
```swift
let connection = try client.connectionPool.checkOut()
defer { 
    if !operation.holdsConnection {
        client.connectionPool.checkIn(connection)
    }
}
return try operation.execute(using: connection, session: nil)
```

Reply:
> I think we probably want to avoid a scenario where certain operations aren't run through the executor.

this is already true for `NextOperation`, which was done so that we avoid checking out a new connection (which would be ignored) each time we call `next`. by manually calling execute we can avoid that. 
if we want those and change stream next ops executed via executor, we might consider adding some flag on an operation that indicates the executor should pass in a dummy connection.

for the operation that calls `watch`, what you propose seems reasonable, though I wonder if we can come up with some abstraction that covers both that and the next operations 🤔

I updated [SWIFT-523](https://jira.mongodb.org/browse/SWIFT-523) with some details from this convo, feel free to edit/add 

ugh, why does mongoc make everything so hard 🤦‍♀

Reply:
We could possibly use the same computed property to determine if we need to check out a connection at all. So the property could be more answering the question "is this operation currently holding onto a connection?" rather than "does this type of operation hold onto its connections?". 

Then we'd have:
- Pre execution: If it is holding onto a connection, don't give it a new one. 
- Post execution: If we checked something out and the operation is still holding onto a connection, don't check anything back in.

Reply:
I agree that operations should pertain to I/O, in particular with a server. A change stream is interesting because its an aggregate operation, and then it becomes a cursor. I like the second idea @patrickfreed, I don't think this is a question so much of whether we should check out a connection, but whether the "execution context" already has a bound connection bypassing the need to check one out at all. Let's defer this to SWIFT-253, but prioritize doing that work

Reply:
ok i've added these comments to the ticket so we can refer back to them later 👍